app/controllers/nih_tables_controller.rb

```diff
@@ -18,119 +16,27 @@ def create_zip_file(samples_report)
   end
 
-  def add_general_samples_table(samples_report, zip_file)
-    csv_file = Tempfile.new("#{samples_report.name}_samples.csv")
-    csv_data = []
-  
-    csv_data << [
-      'sample_id',
-      'vqa_box_id',
-      'sample_group',
-      'sample_media',
-      'sample_media_id',
-      'sample_media_source',
-      'sample_media_source_url',
-      'sample_concentration',
-      'sample_concentration_unit',
-      'sample_concentration_reference_gene',
-      'virus_sample_id',
-      'virus_batch_number',
-      'target_analyte_type',
-      'target_organism_name',
-      'target_organism_taxonomy_id',
-      'pango_lineage',
-      'who_label',
-      'virus_sample_inactivation_method'
-    ]
-  
-    samples_report.samples.each do |sample|
-      batch = Batch.find_by(batch_number: sample.batch_number)
-      csv_data << [
-        sample.id,
-        sample.box.uuid,
-        "#{sample.box.purpose}-panel",
-        nil,
-        nil,
-        nil,
-        nil,
-        sample.concentration,
-        'copies/ml',
-        batch.reference_gene,
-        sample.id,
-        sample.batch_number,
-        'inactivated virus',
-        'SARS-CoV-2',
-        batch.target_organism_taxonomy_id,
-        batch.pango_lineage,
-        batch.who_label,
-        sample.inactivation_method
-      ]
-    end
-  
-    CSV.open(csv_file.path, 'w') do |csv|
-      csv_data.each do |row|
-        csv << row
-      end
-    end
-  
-    zip_file.add("#{samples_report.name}_samples.csv", csv_file.path)
-  end
-  
-  def add_general_results_table(samples_report, zip_file)
-    csv_file = Tempfile.new("#{samples_report.name}_results.csv")
-    csv_data = []
-  
-    csv_data << [
-      'sample_id',
-      'vqa_box_id',
-      'sample_group',
-      'protocol_id',
-      'technology_platform',
-      'assay_readout',
-      'assay_readout_unit',
-      'assay_readout_description'
-    ]
-  
-    samples_report.samples.each do |sample|
-      csv_data << [
-        sample.id,
-        sample.box.uuid,
-        "#{sample.box.purpose}-panel",
-        nil,
-        nil,
-        sample.measured_signal,
-        nil,
-        nil
-      ]
-    end
-  
-    CSV.open(csv_file.path, 'w') do |csv|
-      csv_data.each do |row|
-        csv << row
-      end
-    end
-  
-    zip_file.add("#{samples_report.name}_results.csv", csv_file.path)
-  end
+  def add_nih_table(samples_report, table_name, zip_file)
+    csv_file = Tempfile.new("#{samples_report.name}_#{table_name}.csv")
+    
+    csv_file.write(
+      render_to_string(:file => 'samples_reports/nih_'+table_name+'.csv.csvbuilder', :locals => { :samples_report => samples_report })
+    )
+    csv_file.close
 
-  def add_lod_table(samples_report, zip_file)
-    # TODO: Add the LOD table to the zip file
-  end
-  
-  def add_challenge_table(samples_report, zip_file)
-    # TODO: Add the Challenge table to the zip file
+    zip_file.add("#{samples_report.name}_#{table_name}.csv", csv_file.path)
   end
```

Comment:
You'll want to delete (unlink) the tempfile after it has been added to the ZIP file.

Reply:
I've tried this myself before (delete and unlink) but I get:

![image](https://github.com/instedd/cdx/assets/13782680/931d8e74-0b06-4389-951d-4943dec51cfa)

if I do 

```
    csv_file.close
    zip_file.add("#{@samples_report.name}_#{table_name}.csv", csv_file.path)
    csv_file.unlink
```

I'm doing it wrongly?

Reply:
I saw that there is an alternative [using streams](https://stackoverflow.com/questions/27577838/using-tempfile-to-create-a-zip-file-in-rails) but also, from the docs:

> When a Tempfile object is garbage collected, or when the Ruby interpreter exits, its associated temporary file is automatically deleted.

How do we should proceed here?

Reply:
Argh, I hoped the Zip builder would generate the Zip file as we add files to it (stream by default), but no, it keeps a list of files to add and eventually generates the Zip file by reading the files to add. Damn.

You may want to try streaming. That may end up simpler as we could skip tempfiles... unless we expect larger CSV, like thousands of samples?

At worst, we're leaking file descriptors, and Ruby will run a GC collect when there are none available, but let's try to avoid that.

Reply:
Same for the tempfile of the generated Zip, it should be deleted after being sent to the user.